src/hosts.rs

```diff
@@ -0,0 +1,323 @@
+    ///
+    /// # Returns
+    ///
+    /// true if the host is ready, false otherwise.
+    pub fn check_readyset_is_ready(&mut self) -> Result<bool, mysql::Error> {
+        match &mut self.conn {
+            Some(conn) => {
+                let rows: Vec<(String, String)> =
+                    conn.query("SHOW READYSET STATUS").unwrap_or(vec![]);
+                for (field, value) in rows {
+                    if field == "Snapshot Status" {
+                        return Ok(value == "Completed");
+                    }
+                }
+                Ok(false)
+            }
+            None => Err(mysql::Error::IoError(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Connection to Readyset host is not established",
+            ))),
+        }
+    }
+
+    /// Checks if the host supports the given query.
+    /// This is done by querying the EXPLAIN CREATE CACHE FROM command.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest_text` - The digest text of the query.
+    /// * `schema` - The schema of the query.
+    ///
+    /// # Returns
+    ///
+    /// true if the host supports the query, false otherwise.
+    pub fn check_query_support(
+        &mut self,
+        digest_text: &String,
+        schema: &String,
+    ) -> Result<bool, mysql::Error> {
+        match &mut self.conn {
+            Some(conn) => {
+                conn.query_drop(format!("USE {}", schema))
+                    .expect("Failed to use schema");
+                let row: Option<(String, String, String)> =
+                    conn.query_first(format!("EXPLAIN CREATE CACHE FROM {}", digest_text))?;
+                match row {
+                    Some((_, _, value)) => Ok(value == "yes" || value == "cached"),
+                    None => Ok(false),
+                }
+            }
+            None => Ok(false),
+        }
+    }
+
+    /// Caches the given query on the host.
+    /// This is done by executing the CREATE CACHE FROM command.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest_text` - The digest text of the query.
+    ///
+    /// # Returns
+    ///
+    /// true if the query was cached successfully, false otherwise.
+    pub fn cache_query(
+        &mut self,
+        digest_text: &String,
+        digest: &String,
+    ) -> Result<bool, mysql::Error> {
+        match &mut self.conn {
+            None => return Ok(false),
+            Some(conn) => {
+                conn.query_drop(format!("CREATE CACHE d_{} FROM {}", digest, digest_text))
+                    .expect("Failed to create readyset cache");
+            }
+        }
+        Ok(true)
+    }
+
+    /// Changes the status of the host in the ProxySQL mysql_servers table.
+    /// The status is set to the given `status`.
+    pub fn change_status(
+        &mut self,
+        ps_conn: &mut Conn,
+        config: &Config,
+        status: HostStatus,
+    ) -> Result<bool, mysql::Error> {
```

Comment:
Doesn't look like we return anything other than `true` so we should probably return `Result<(), ...>`

Reply:
This function migrated to `ProxySQL::health_check` - ProxySQL handles all the changes required to proxysql and host.change_status is just an in memory version of the state of the host.